pageserver/src/tenant/layer_map.rs

```diff
@@ -274,17 +274,17 @@ where
     ///
     /// Helper function for BatchedUpdates::insert_historic
     ///
-    pub(self) fn insert_historic_noflush(&mut self, layer: Arc<L>) {
-        self.historic.insert(
-            historic_layer_coverage::LayerKey::from(&*layer),
-            Arc::clone(&layer),
-        );
+    pub(self) fn insert_historic_noflush(&mut self, layer: Arc<L>) -> bool {
+        let key = historic_layer_coverage::LayerKey::from(&*layer);
+        let exists = self.historic.contains(&key);
+        self.historic.insert(key, Arc::clone(&layer));
 
         if Self::is_l0(&layer) {
             self.l0_delta_layers.push(layer);
         }
 
         NUM_ONDISK_LAYERS.inc();
+        !exists
```

Comment:
```suggestion
        if self.historic.contains(&key) {
            return false;
        }
        self.historic.insert(key, Arc::clone(&layer));

        if Self::is_l0(&layer) {
            self.l0_delta_layers.push(layer);
        }

        NUM_ONDISK_LAYERS.inc();
        true
```

Note that otherwise for example the `NUM_ONDISK_LAYERS` would be wrong. Though it is completly wrong right now by other reasons, I hope to continue with #3775 some day.

Reply:
Yes, I agree with you.I rewrite this function - it is not returning book any more, duplicates are checked using assert

Reply:
Ok, thanks for responding. We have another thread for `assert!` so I'll resolve this in favor of that: https://github.com/neondatabase/neon/pull/3869#discussion_r1147285293